### PR TITLE
Provision Plugins

### PR DESCRIPTION
**Pull request by sdan**

### CODE REVIEW

File: gateway/plugin.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	g.authorize(ctx, gqlReq.Header)
+
+	gqlReq.Var("projectId", req.ProjectID)
+	gqlReq.Var("name", req.Plugin)
+
+	var resp struct {
+		Plugin *entity.Plugin `json:"createProject"`
+	}
+	if err := g.gqlClient.Run(ctx, gqlReq, &resp); err != nil {
+		//TODO: delete this line
```
Comment (JakeCooper) [inline gateway/plugin.go:31]: todo

File: cmd/add.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	"strings"
+
+	"github.com/railwayapp/cli/entity"
+	"github.com/railwayapp/cli/errors"
+	"github.com/railwayapp/cli/ui"
+)
+
+func (h *Handler) Add(ctx context.Context, req *entity.CommandRequest) error {
+	if len(req.Args) == 0 {
+		return errors.PluginNotSpecified
```
Comment (JakeCooper) [inline cmd/add.go:15]: Do prompt here for plugins

File: cmd/add.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		return errors.PluginAlreadyExists
+	}
+	_, err = h.ctrl.CreatePlugin(ctx, &entity.CreatePluginRequest{
+		ProjectID: project.Id,
+		Plugin:    pluginRequest,
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Printf("🎉 Created plugin %s\n", ui.MagentaText(pluginRequest))
```
Comment (JakeCooper) [inline cmd/add.go:42]: Should use the plugin response with the plugin.Name or whatever instead of just burying it

File: controller/plugin.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	return c.gtwy.CreatePlugin(ctx, req)
+}
+
+func (c *Controller) GetPlugins(ctx context.Context, projectId string) ([]*entity.Plugin, error) {
+	projectCfg, err := c.gtwy.GetProject(ctx, projectId)
+	if err != nil {
+		return nil, err
+	}
+	plugins := projectCfg.Plugins
+	return plugins, nil
```
Comment (JakeCooper) [inline controller/plugin.go:19]: just return projectCfg.Plugins instead of extra variable here

File: controller/plugin.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+func (c *Controller) GetPlugins(ctx context.Context, projectId string) ([]*entity.Plugin, error) {
+	projectCfg, err := c.gtwy.GetProject(ctx, projectId)
+	if err != nil {
+		return nil, err
+	}
+	return projectCfg.Plugins, nil
+}
+
+func availablePlugins(pluginRequest string) []*entity.Plugin {
+	set := map[string]bool{"env": true, "postgresql": true, "mongodb": true, "redis": true}
```
Comment (JakeCooper) [inline controller/plugin.go:22]: These need to be fetched remotely

File: cmd/add.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+		return err
+	}
+
+	project, err := h.ctrl.GetProject(ctx, projectCfg.Project)
+	if err != nil {
+		return err
+	}
+
+	pluginRequest := strings.TrimSpace(req.Args[0])
+	allowCreation, pluginsAvailable, err := h.ctrl.PluginExists(ctx, pluginRequest, project.Id)
```
Comment (JakeCooper) [inline cmd/add.go:24]: Triple return is almost never the move. Additionally, the plugin request should take the projectID as part of the payload. Make an entity

Suggested abstraction: 
```
[]Plugin, err := h.ctrl.GetAvailablePlugins(ctx, project.Id)
```

File: errors/main.go [hunk 7]
```diff
@@ -18,4 +18,8 @@ var (
 	CommandNotSpecified     RailwayError = errors.New(fmt.Sprintf("%s\nRun %s", ui.RedText("Specify a command to run inside the railway environment."), ui.Bold("railway run <cmd>")))
 	LoginFailed             RailwayError = errors.New(fmt.Sprintf("%s", ui.RedText("Login failed")))
 	LoginTimeout            RailwayError = errors.New(fmt.Sprintf("%s", ui.RedText("Login timeout")))
+	PluginAlreadyExists     RailwayError = errors.New(fmt.Sprintf("%s", ui.RedText("Plugin already exists")))
+	PluginNotSpecified      RailwayError = errors.New(fmt.Sprintf("%s\nRun %s", ui.RedText("Specify a plugin to create."), ui.Bold("railway add <plugin>")))
+	PluginCreateFailed      RailwayError = errors.New(fmt.Sprintf("%s\nOne of our trains probably derailed!", ui.RedText("There was a problem creating the plugin.")))
+	PluginGetFailed         RailwayError = errors.New(fmt.Sprintf("%s\nOne of our trains probably derailed!", ui.RedText("There was a problem getting plugins available for creation.")))
```
Comment (JakeCooper) [inline errors/main.go:24]: Random copy update

```suggestion
	PluginCreateFailed      RailwayError = errors.New(fmt.Sprintf("%s\nUhh Ohh! One of our trains derailed.", ui.RedText("There was a problem creating the plugin.")))
	PluginGetFailed         RailwayError = errors.New(fmt.Sprintf("%s\nUhh Ohh! One of our trains derailed.", ui.RedText("There was a problem getting plugins available for creation.")))
```

File: cmd/add.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    for _, b := range list {
+        if *b == a {
+            return true
+        }
+    }
+    return false
+}
+
+func (h *Handler) addFromName(ctx context.Context, pluginRequest string, plugins []*string, projectId string) error {
+	reqValid := stringInSlice(pluginRequest,plugins)
```
Comment (JakeCooper) [inline cmd/add.go:22]: Space after comma. Is your IDE setup for gopls?

File: cmd/add.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	if !reqValid {
+		return errors.New("Plugin either has been created or not valid.")
+	}
```
Comment (JakeCooper) [inline cmd/add.go:25]: Should add a seperate check for invalid plugin. Honestly, probably just better to only have the dropdown. Args may confuse people and they'll need to know exact string (e.g postgresql vs postgres)

```suggestion
	reqValid := stringInSlice(pluginRequest ,plugins)
	if !reqValid {
		return errors.New("Plugin already exists.")
	}
```

File: gateway/plugin.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+			availablePluginsForProject(projectId: $projectId)
+		}
+	`)
+
+	gqlReq.Var("projectId", projectId)
+
+	g.authorize(ctx, gqlReq.Header)
+
+	var resp struct {
+		Plugins []*string `json:"availablePluginsForProject"`
```
Comment (JakeCooper) [inline gateway/plugin.go:23]: What's the utility of having an array of pointers here vs just an array of strings?

Comment (sdan) [reply gateway/plugin.go:23]: was just trying to follow convention. good point though; ran into some memory issues along the way.

Comment (JakeCooper) [reply gateway/plugin.go:23]: Yea. You'll also just be derefing things for no reason (like in your plugin search function)

File: cmd/add.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/railwayapp/cli/entity"
+	"github.com/railwayapp/cli/ui"
+)
+
+func stringInSlice(a string, list []*string) bool {
+    for _, b := range list {
+        if *b == a {
```
Comment (JakeCooper) [inline cmd/add.go:14]: I know you're likely to replace with an array of strings, but in the future, when you deref something, really should have a nil check. You're asking for panics here

**Review by JakeCooper — APPROVED** [review]

**Review by coffee-cup — APPROVED** [review]

**Pull request merged**